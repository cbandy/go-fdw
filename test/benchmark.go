package main

import "C"
import (
	"fmt"
	"runtime"

	fdw "github.com/cbandy/go-fdw"
)

var benchmark struct {
	mem runtime.MemStats
}

//export goTestBenchmarkBegin
func goTestBenchmarkBegin() {
	runtime.ReadMemStats(&benchmark.mem)
}

//export goTestBenchmarkEnd
func goTestBenchmarkEnd(dest []byte) C.int {
	var end runtime.MemStats
	runtime.ReadMemStats(&end)

	s := fmt.Sprintf("%8d B\t%8d allocs",
		end.TotalAlloc-benchmark.mem.TotalAlloc,
		end.Mallocs-benchmark.mem.Mallocs)

	return C.int(copy(dest, s))
}

type benchmarkPath string
type benchmarkIterator struct {
	data [][]string
	row  int
}

var benchmarkConst = [][]string{
	{"one", "a", "b", "Lorem ipsum dolor sit amet"},
	{"two", "c", "d", "consectetur adipiscing elit"},
	{"three", "e", "f", "sed do eiusmod tempor incididunt"},
	{"four", "g", "h", "ut labore et dolore magna aliqua"},
}

func (p benchmarkPath) Estimate(cost fdw.ScanCostEstimate) (fdw.ScanCostEstimate, error) {
	return cost, nil
}

func (p benchmarkPath) Begin() (fdw.Iterator, error) {
	return &benchmarkIterator{data: benchmarkConst}, nil
}

func (p benchmarkPath) Close() error { return nil }

func (i *benchmarkIterator) Next(attr []fdw.Attribute) (bool, error) {
	if i.row < len(i.data) {
		for j, a := range attr {
			a.SetString(i.data[i.row][j])
		}
		i.row++
		return true, nil
	}
	return false, nil
}

func (benchmarkIterator) Close() error { return nil }
