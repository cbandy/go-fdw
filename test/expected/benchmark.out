-- vim: set expandtab shiftwidth=2 syntax=pgsql tabstop=2 :
--
DO $$ BEGIN RAISE INFO 'benchmark'; END $$;
INFO:  benchmark
--
CREATE SERVER benchmark FOREIGN DATA WRAPPER go_test_fdw
OPTIONS (test 'benchmark');
--
CREATE FOREIGN TABLE one_column (col1 text) SERVER benchmark;
DO $$ BEGIN PERFORM * FROM one_column; END; $$;
DO $$ BEGIN PERFORM go_test_fdw_benchmark_begin(); END; $$;
DO $$ BEGIN PERFORM * FROM one_column; END; $$;
SELECT go_test_fdw_benchmark_end();
    go_test_fdw_benchmark_end    
---------------------------------
     1432 B            40 allocs
(1 row)

--
CREATE FOREIGN TABLE four_columns (col1 text, col2 text, col3 text, col4 text) SERVER benchmark;
DO $$ BEGIN PERFORM * FROM four_columns; END; $$;
DO $$ BEGIN PERFORM go_test_fdw_benchmark_begin(); END; $$;
DO $$ BEGIN PERFORM * FROM four_columns; END; $$;
SELECT go_test_fdw_benchmark_end();
    go_test_fdw_benchmark_end    
---------------------------------
     1608 B            44 allocs
(1 row)

--
-- cleanup
--
DROP SERVER benchmark CASCADE;
NOTICE:  drop cascades to 2 other objects
DETAIL:  drop cascades to foreign table one_column
drop cascades to foreign table four_columns
